code/streaming/Hacl.Streaming.Keccak.fst

```diff
@@ -0,0 +1,242 @@
+let uint64 = Lib.IntTypes.uint64
+
+open Spec.Hash.Definitions
+
+open Hacl.Streaming.Interface
+
+open Hacl.Hash.Definitions
+
+module D = Hacl.Hash.Definitions
+module Agile = Spec.Agile.Hash
+
+let _: squash (inversion hash_alg) = allow_inversion hash_alg
+
+/// This is a dedicated streaming API for the Keccak construction. The reason we
+/// are not piggybacking on Hacl.Streaming.MD is that we want a *single* piece
+/// of code that can deal with all 6 variants at runtime, as opposed to a
+/// meta-programmed version that requires six instantiations each for each one
+/// of the Keccak algorithms. Notably, this means that we keep the rate at
+/// run-time (relying on the key), and that we request the outputByteLen for the
+/// finish function.
+
+inline_for_extraction noextract
+let alg = sha3_alg
+
+unfold let coerce (#b #a:Type) (x:a{a == b}) : b = x
+
+inline_for_extraction noextract
+let update_multi_s (a : alg) acc (prevlen : nat) input =
+  Agile.update_multi a acc () input
+
+noextract
+let update_multi_zero (a : alg) acc (prevlen : nat) :
+  Lemma(update_multi_s a acc prevlen S.empty == acc) = Spec.Hash.Lemmas.update_multi_zero a acc
+
+// TODO: this is the fourth copy of this lemma!! why?!
+#push-options "--ifuel 1"
+noextract
+let update_multi_associative (a : alg) acc (prevlen1 prevlen2 : nat)
+                             (input1 input2 : S.seq uint8) :
+    Lemma
+    (requires (
+      prevlen1 % U32.v (D.block_len a) = 0 /\
+      S.length input1 % U32.v (D.block_len a) = 0 /\
+      S.length input2 % U32.v (D.block_len a) = 0 /\
+      prevlen2 = prevlen1 + S.length input1))
+    (ensures (
+      let input = S.append input1 input2 in
+      S.length input % U32.v (D.block_len a) = 0 /\
+      prevlen2 % U32.v (D.block_len a) = 0 /\
+      update_multi_s a (update_multi_s a acc prevlen1 input1) prevlen2 input2 ==
+        update_multi_s a acc prevlen1 input)) =
+  Spec.Hash.Lemmas.update_multi_associative a acc input1 input2
+#pop-options
+
+let singleton #t (x: t) = y:t { y == x }
+
+inline_for_extraction noextract
+let stateful_keccak: stateful alg =
+  Stateful
+    (* s: *) (fun (a: alg) ->
+      singleton a & b:B.buffer uint64 { B.len b == 25ul })
+    (* footprint: *) (fun #_ h (_, s) ->
+      B.loc_addr_of_buffer s)
+    (* freeable: *) (fun #_ h (_, s) ->
+      B.freeable s)
+    (* invariant: *) (fun #_ h (_, s) ->
+      B.live h s)
+    (* t: *) (fun _ ->
+      s:S.seq uint64 { S.length s == 25 })
+    (* v: *) (fun _ h (a, s) ->
+       B.as_seq h s)
+    (* invariant_loc_in_footprint: *) (fun #_ h (_, s) ->
+      ())
+    (* frame_invariant: *) (fun #_ l (_, s) h0 h1 ->
+      ())
+    (* frame_freeable: *) (fun #_ l (_, s) h0 h1 ->
+      ())
+    (* alloca: *) (fun (a: alg) ->
+      a, B.alloca (Lib.IntTypes.u64 0) 25ul)
+    (* create_in: *) (fun a r ->
+      a, B.malloc r (Lib.IntTypes.u64 0) 25ul)
+    (* free: *) (fun _ (_, s) ->
+      B.free s)
+    (* copy: *) (fun _ (a, s_src) (a', s_dst) ->
```

Comment:
this is where the singleton type is needed: because the two types share the same index (as imposed by the type signature of copy), we can conclude that the two values are exactly the same, meaning that a and a' are identical